File: src/baekjoon/bfs/BOJ_1325.java
package baekjoon.bfs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class BOJ_1325 {

    static ArrayList<ArrayList<Integer>> list;
    static boolean[] visited;
    static int[] answer;
    public static void main(String[] args) throws IOException {


        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken()); // 컴퓨터 수
        int M = Integer.parseInt(st.nextToken()); // 신뢰 관계 수

        answer = new int[N+1];
        list = new ArrayList<>();

        for (int i=0; i<=N; i++) list.add(new ArrayList<>());

        // A가 B를 신뢰하는 경우 -> B를 해킹하면, A도 해킹 가능
        // input A B -> A가 B를 신뢰.
        for (int i=0; i<M; i++){
            st = new StringTokenizer(br.readLine());

            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            list.get(a).add(b); // b 를 신뢰하는 list[a]
        }

        for (int i=1; i<=N; i++){
            visited = new boolean[N+1];
            bfs(i);
        }

        int max = Integer.MIN_VALUE;

        // 최댓값
        for (int i=1; i<=N; i++){
            max = Math.max(max, answer[i]);
        }

        for (int i=1; i<=N; i++){
            if (answer[i]==max){
                System.out.print(i + " ");
            }
        }

    }

    static void bfs(int start){
        Queue<Integer> queue = new LinkedList<>();

        queue.add(start);
        visited[start] = true;

        while (!queue.isEmpty()){
            int now = queue.poll();

            // now -> num
            // now가 num을 신뢰함
            for (int num : list.get(now)){
                if (!visited[num]){
                    visited[num] = true;

                    answer[num]++;
                    queue.add(num);
                }
            }
        }
    }

}

File: src/baekjoon/bfs/BOJ_1325.kt
package baekjoon.bfs

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

class BOJ_1325 {///

    private lateinit var graph: Array<MutableList<Int>>
    private lateinit var visited: BooleanArray

    fun solve() {
        val br = BufferedReader(InputStreamReader(System.`in`))
        val (N, M) = br.readLine().split(" ").map { it.toInt() } // 컴퓨터 수 N 관계 수 M

        graph = Array(N + 1) { mutableListOf() }

        repeat(M) {
            val (A, B) = br.readLine().split(" ").map { it.toInt() }
            graph[B].add(A) // B가 해킹 -> A도 해킹
        }

        val result = IntArray(N + 1) // 해킹 가능한 수 d
        var maxCnt = 0 // 최대 해킹 가능 수

        for (i in 1..N) {
            visited = BooleanArray(N + 1)
            val cnt = bfs(i)
            result[i] = cnt
            maxCnt = maxOf(maxCnt, cnt)
        }

        val answer = mutableListOf<Int>()
        for (i in 1..N) {
            if (result[i] == maxCnt) { // 최대 해킹 가능한 컴퓨터 번호
                answer.add(i)
            }
        }

        println(answer.joinToString(" "))
    }

    private fun bfs(start: Int): Int {
        val queue: Queue<Int> = LinkedList()
        queue.add(start)
        visited[start] = true

        var count = 1 // 시작 노드

        while (queue.isNotEmpty()) {
            val current = queue.poll()

            for (next in graph[current]) {
                if (!visited[next]) {
                    visited[next] = true
                    queue.add(next)
                    count++
                }
            }
        }

        return count // 총 해킹 가능 컴퓨터 수
    }
}

fun main() {
    BOJ_1325().solve()
}